File: src/test/java/com/github/kayjamlang/tests/expressions/ifExpr/IfWithExpressionTest.java
package com.github.kayjamlang.tests.expressions.ifExpr;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.IfExpression;
import com.github.kayjamlang.core.expressions.VariableLinkExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class IfWithExpressionTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("if(putin) true else false"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(IfExpression.class, expression.getClass());

        IfExpression ifExpression = (IfExpression) expression;
        assertSame(VariableLinkExpression.class, ifExpression.condition.getClass());

        assertNotNull(ifExpression.ifTrue);
        assertNotNull(ifExpression.ifFalse);

        assertSame(ValueExpression.class, ifExpression.ifTrue.getClass());
        assertSame(ValueExpression.class, ifExpression.ifFalse.getClass());

        VariableLinkExpression root = (VariableLinkExpression) ifExpression.condition;
        assertEquals("putin", root.name);

        ValueExpression ifTrue = (ValueExpression) ifExpression.ifTrue;
        assertEquals(true, ifTrue.value);

        ValueExpression ifFalse = (ValueExpression) ifExpression.ifFalse;
        assertEquals(false, ifFalse.value);
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/ifExpr/IfWithExpressionTest.scala
package com.github.kayjamlang.tests.expressions.ifExpr

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{IfExpression, ValueExpression, VariableLinkExpression}
import org.junit.Assert._
import org.junit.{BeforeClass, Test}


object IfWithExpressionTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer(
      """
        |if (putin)
        | true
        |else
        | false
        |""".stripMargin))
  }
}

class IfWithExpressionTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = IfWithExpressionTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[IfExpression], expression getClass)
    val ifExpression = expression.asInstanceOf[IfExpression]
    assertSame(classOf[VariableLinkExpression], ifExpression.condition getClass)
    assertNotNull(ifExpression.ifTrue)
    assertNotNull(ifExpression.ifFalse)
    assertSame(classOf[ValueExpression], ifExpression.ifTrue getClass)
    assertSame(classOf[ValueExpression], ifExpression.ifFalse getClass)
    val root = ifExpression.condition.asInstanceOf[VariableLinkExpression]
    assertEquals("putin", root.name)
    val ifTrue = ifExpression.ifTrue.asInstanceOf[ValueExpression]
    assertEquals(true, ifTrue.value)
    val ifFalse = ifExpression.ifFalse.asInstanceOf[ValueExpression]
    assertEquals(false, ifFalse.value)
  }
}